compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ),
                 Mismatch::Fixed(s) => (s.into(), s.into(), None),
             };
-            match (&terr, expected == found) {
-                (TypeError::Sorts(values), extra) => {
+            let looks_similar = |e: ExpectedFound<Ty<'_>>| {
+                // We're only interested in adts
+                if let (Some(e), Some(f)) = (e.expected.ty_adt_def(), e.found.ty_adt_def()) {
+                    // Only compare the last parts of the path.
+                    // `whatever::Foo` is pretty similar to `blah::Foo`
+                    let e_path = self.tcx.def_path(e.did()).data;
+                    let f_path = self.tcx.def_path(f.did()).data;
+                    if let (Some(e), Some(f)) = (e_path.last(), f_path.last()) {
+                        return e.data == f.data;
+                    }
+                }
+                false
+            };
+
+            match terr {
+                // If two types mismatch but have similar names, mention that specifically.
+                TypeError::Sorts(values) if looks_similar(values) => {
+                    let found_adt = values.found.ty_adt_def().unwrap();
+                    let expected_adt = values.expected.ty_adt_def().unwrap();
+
+                    let found_name = values.found.sort_string(self.tcx);
+                    let expected_name = values.expected.sort_string(self.tcx);
+
+                    diag.note(format!("{found_name} and {expected_name} have similar names, but are actually distinct types"));
```

Comment:
how does this interact with the existing "help: maybe two different crates are being used?" message? IMO we should probably only show one or the other; it would be nice to add a test case.

Reply:
It shows both; see `src/test/ui/type/type-mismatch-same-crate-name`. Its output got updated with this PR.